File: Singleton/java/RegistrySingleton.java
// 注册表单例
import java.util.HashMap;
import java.util.Map;

public class RegistrySingleton {

  private RegistrySingleton() {}

  private static Map registry = new HashMap<String, Object>();

  public static RegistrySingleton getInstance(String classname){
    Object singleton = registry.get(classname);
    if (singleton == null) {
      try {
        singleton = Class.forName(classname).newInstance();
      } catch (Exception e) {
        e.printStackTrace();
      }
      registry.put(classname,singleton);
    }
    return (RegistrySingleton) singleton;
  }

  public void check() {
    System.out.println("the instance has been created");
  }

  public static void main(String[] args) {
    getInstance("RegistrySingleton").check();
  }
}

File: singleton/python/RegistrySingleton.py
#!/usr/bin/env python
# coding:utf8


class RegistrySingleton(object):
    registry = {}

    @classmethod
    def getSingleton(cls):
        if cls not in cls.registry:
            RegistrySingleton.registry[cls] = object.__new__(cls)
        return cls.registry[cls]

    def check(self):
        print("the instance has been created")

    def __init__(self):
        raise "createError"


if __name__ == "__main__":
    RegistrySingleton.getSingleton().check()

# 不能通过 m n o 测试的原因是，这里没有限制实例化，所以实例化和注册表是两种不同的实例化方式。
